File: Packages/OverlayKit/Sources/OverlayKit/PinGeometry.swift
import CoreGraphics

/// Pure geometry for pinned screenshots (all rects in Cocoa global points).
public enum PinGeometry {
    /// Initial pin frame: the image's point size (pixels ÷ backingScale),
    /// shrunk if needed so it fits within maxFraction of the visible frame
    /// (aspect preserved), centered on sourceRect when known (else on the
    /// visible frame), then nudged fully on-screen.
    public static func initialFrame(imagePixelSize: CGSize, backingScale: CGFloat,
                                    visibleFrame: CGRect, sourceRect: CGRect?,
                                    maxFraction: CGFloat = 0.8) -> CGRect {
        var size = CGSize(width: imagePixelSize.width / backingScale,
                          height: imagePixelSize.height / backingScale)
        let scale = min(1, (visibleFrame.width * maxFraction) / size.width,
                        (visibleFrame.height * maxFraction) / size.height)
        size = CGSize(width: size.width * scale, height: size.height * scale)

        let center = sourceRect.map { CGPoint(x: $0.midX, y: $0.midY) }
            ?? CGPoint(x: visibleFrame.midX, y: visibleFrame.midY)
        var frame = CGRect(x: center.x - size.width / 2, y: center.y - size.height / 2,
                           width: size.width, height: size.height)
        if frame.maxX > visibleFrame.maxX { frame.origin.x = visibleFrame.maxX - frame.width }
        if frame.maxY > visibleFrame.maxY { frame.origin.y = visibleFrame.maxY - frame.height }
        if frame.minX < visibleFrame.minX { frame.origin.x = visibleFrame.minX }
        if frame.minY < visibleFrame.minY { frame.origin.y = visibleFrame.minY }
        return frame
    }

    /// Rescales `current` around its center by `factor`, clamped so the result
    /// stays between minScale× and maxScale× of `naturalSize` (aspect preserved).
    public static func zoomedFrame(current: CGRect, naturalSize: CGSize, factor: CGFloat,
                                   minScale: CGFloat = 0.25, maxScale: CGFloat = 3.0) -> CGRect {
        guard naturalSize.width > 0, naturalSize.height > 0, factor > 0 else { return current }
        let newScale = min(maxScale, max(minScale, (current.width / naturalSize.width) * factor))
        let newSize = CGSize(width: naturalSize.width * newScale,
                             height: naturalSize.height * newScale)
        return CGRect(x: current.midX - newSize.width / 2,
                      y: current.midY - newSize.height / 2,
                      width: newSize.width, height: newSize.height)
    }
}

File: windows/src/BetterScreenshot.Capture/PinGeometry.cs
using BetterScreenshot.Core;

namespace BetterScreenshot.Capture;

/// <summary>
/// Pure geometry for pin-to-screen panels: the initial floating frame for a captured image (sized in logical
/// units, clamped to ≤80% of the visible area, centered on the source region and kept on-screen), and
/// aspect-locked zoom about the frame center clamped to 0.25×–3×.
/// </summary>
public static class PinGeometry
{
    public static PxRect InitialFrame(PxSize imagePixelSize, double dpiScale, PxRect visibleFrame, PxRect? sourceRect = null, double maxFraction = 0.8)
    {
        double w = imagePixelSize.Width / dpiScale;
        double h = imagePixelSize.Height / dpiScale;

        // Shrink (never enlarge) to fit within maxFraction of the visible area, preserving aspect.
        double maxW = visibleFrame.Width * maxFraction;
        double maxH = visibleFrame.Height * maxFraction;
        double scale = Math.Min(1.0, Math.Min(maxW / w, maxH / h));
        w *= scale;
        h *= scale;

        PxPoint center = sourceRect?.Center ?? visibleFrame.Center;
        double x = center.X - w / 2;
        double y = center.Y - h / 2;

        // Keep fully inside the visible frame.
        x = Math.Clamp(x, visibleFrame.X, Math.Max(visibleFrame.X, visibleFrame.Right - w));
        y = Math.Clamp(y, visibleFrame.Y, Math.Max(visibleFrame.Y, visibleFrame.Bottom - h));
        return new PxRect(x, y, w, h);
    }

    public static PxRect ZoomedFrame(PxRect current, PxSize naturalSize, double factor, double minScale = 0.25, double maxScale = 3.0)
    {
        double currentScale = current.Width / naturalSize.Width;
        double newScale = Math.Clamp(currentScale * factor, minScale, maxScale);
        double newW = naturalSize.Width * newScale;
        double newH = naturalSize.Height * newScale;
        var c = current.Center;
        return new PxRect(c.X - newW / 2, c.Y - newH / 2, newW, newH);
    }
}
